src/libfat32.s

```diff
@@ -1,629 +1,1390 @@
+  ;
+  ; Otherwise, A is set to the file's attribute byte and
+  ; zp_sd_address points at the returned directory entry.
+  ; LFNs and empty entries are ignored automatically.
+
+  ; Increment pointer by 32 to point to next entry
+  clc
+  lda zp_sd_address
+  adc #32
+  sta zp_sd_address
+  lda zp_sd_address+1
+  adc #0
+  sta zp_sd_address+1
+
+  ; If it's not at the end of the buffer, we have data already
+  cmp #>(fat32_readbuffer+$200)
+  bcc .gotdata
+
+  ; Read another sector
+  lda #<fat32_readbuffer
+  sta fat32_address
+  lda #>fat32_readbuffer
+  sta fat32_address+1
+
+  jsr fat32_readnextsector
+  bcc .gotdata
+
+.endofdirectory
+  sec
+  rts
+
+.gotdata
+  ; Check first character
+  ldy #0
+  lda (zp_sd_address),y
+
+  ; End of directory => abort
+  beq .endofdirectory
+
+  ; Empty entry => start again
+  cmp #$e5
+  beq fat32_readdirent
+
+  ; Check attributes
+  ldy #11
+  lda (zp_sd_address),y
+  and #$3f
+  cmp #$0f ; LFN => start again
+  beq fat32_readdirent
+
+  ; Yield this result
+  clc
+  rts
+
+
+fat32_finddirent:
+  ; Finds a particular directory entry. X,Y point to the 11-character filename to seek.
+  ; The directory should already be open for iteration.
+
+  ; Form ZP pointer to user's filename
+  stx fat32_filenamepointer
+  sty fat32_filenamepointer+1
+  
+  ; Iterate until name is found or end of directory
+.direntloop
+  jsr fat32_readdirent
+  ldy #10
+  bcc .comparenameloop
+  rts ; with carry set
+
+.comparenameloop
+  lda (zp_sd_address),y
+  cmp (fat32_filenamepointer),y
+  bne .direntloop ; no match
+  dey
+  bpl .comparenameloop
+
+  ; Found it
+  clc
+  rts
+
+fat32_markdeleted:
+  ; Mark the file as deleted
+  ; We need to stash the first character at index 0x0D
+  ldy #$00
+  lda (zp_sd_address),y
+  ldy #$0d
+  sta (zp_sd_address),y
+
+  ; Now put 0xE5 at the first byte
+  ldy #$00
+  lda #$e5 
+  sta (zp_sd_address),y
+
+  ; Get start cluster high word
+  ldy #$14
+  lda (zp_sd_address),y
+  sta fat32_nextcluster+2
+  iny
+  lda (zp_sd_address),y
+  sta fat32_nextcluster+3
+
+  ; And low word
+  ldy #$1a
+  lda (zp_sd_address),y
+  sta fat32_nextcluster
+  iny
+  lda (zp_sd_address),y
+  sta fat32_nextcluster+1
+
+  ; Write the dirent
+  jsr fat32_wrcurrent
+
+  ; Done
+  clc
+  rts
+
+fat32_deletefile:
+  ; Removes the open file from the SD card.
+  ; The directory needs to be open and
+  ; zp_sd_address pointed to the first byte of the file entry.
+
+  ; Mark the file as "Removed"
+  jsr fat32_markdeleted
+
+  ; We will read a new sector the first time around
+  stz fat32_lastsector
+  stz fat32_lastsector+1
+  stz fat32_lastsector+2
+  stz fat32_lastsector+3
+
+  ; Now we need to iterate through this file's cluster chain, and remove it from the FAT.
+  ldy #0
+.chainloop
+  ; Seek to cluster
+  sec
+  jsr fat32_seekcluster
+
+  ; Is this the end of the chain?
+  lda fat32_nextcluster+3
+  bmi .endofchain
+
+  ; Zero it out
+  lda #0
+  sta (zp_sd_address),y

```

Comment:
assuming that the idea here is to zero the entry in the FAT for the cluster being freed.  And then to walk the cluster chain and do the same for each fat entry.  I think Y register following the call to fat32_seekcluster will not be the entry for the current cluster, but for the current cluster+1,  doesn't Y need to be decremented before zeroing out?  Looking at fat32_seekcluster, it seems the last operations with Y are to increment through the 4 bytes of the current cluster.  Am I missing something here?

Reply:
huh, very true. i haven't tested it yet, but i have a commit that fixes this (i think)

Reply:
i'll put my test results here once i rebased these changes to XPL-32

Reply:
huh, strangely, the XPL-32 copy of this repo already had this patch... exept for the `iny` lines.

i'll test this now.

Reply:
actually, we don't need to iny. y is overwritten by `fat32_seekcluster`.